Fix the "Transfer to another user" screen in the hosting dashboard

markbiek opened the pull request:
<!--
Link a related Github/Linear issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword. Consider using "part of" instead.
-->

Fixes ARC-1294

## Proposed Changes

**Scenario 1**
A CIAB user is attempting to transfer a domain mapping to another user on the same CIAB site. In this case, we only see a "-" since no display name is available and the user id is directly from the CIAB site and is not a WPCOM user id.

<img width="1612" height="1292" alt="image" src="https://github.com/user-attachments/assets/05b9026a-3990-4771-8563-74f72522956c" />

**Scenario 2**
A WPCOM user (with a WoA) site is attempting to transfer a domain mapping to another user on the site. The other user was manually created on the site, then attached to a WPCOM account later. In this case, we see the correct display name for the other user but the user id is still directly from the site and is not a WPCOM user id.

**The fix**
Previously, we called `/wp/v2/:siteId/users` to populate the "New Owner" pulldown. This (generally) went directly to the remote site.

Now we call the `/rest/v1.1/sites/:siteId/users?force=wpcom` endpoint. This ensures that we always get the correct WPCOM user data, even in cases like CIAB.

<img width="759" height="654" alt="image" src="https://github.com/user-attachments/assets/5a1e44d3-efba-4837-9614-3bdbef9d80cf" />


## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Create a new CIAB site and invite another user to the site.
* Attach an external domain to the site (no need to setup DNS).
* Run this locally or use the calypso.live url and visit `/ciab/domains/the.external-domain.com/transfer/other-user`.
* Verify that the pulldown contains the correct display name and user id for the other user.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

markbiek committed:
New fetcher to call /sites/:siteId/users on WPCOM

By default, we force=wpcom to get the user data from WPCOM instead of
the site itself

markbiek committed:
Fix transfer-domain-to-user pulldown

Previously, we called /wp/v2/:siteId/users which returned the user id
from the site itself (which often wasn't the WPCOM user id).

Now we use the new fetcher to call /rest/v1.1/sites/:siteId/users to get
the user data from WPCOM.

p-jackson commented on packages/api-queries/src/site-users.ts:
```diff
@@ -24,3 +29,9 @@ export const siteUsersQuery = ( siteId: number ) =>
+
+export const siteUsersWpcomQuery = ( siteId: number, role?: string ) =>
+	queryOptions( {
+		queryKey: [ 'site', siteId, 'users', 'wpcom', role ],
```
I'd go with a query key like

```suggestion
		queryKey: [ 'site', siteId, 'wpcom-users', role ],
```

It looks like "users" and "wpcom users" are different entities, and so we wouldn't want something like `queryClient.invalidateQueries( { queryKey: [ 'site', siteId, 'users' ] } )` to also invalidate the wpcom users, since they're unrelated.

p-jackson commented on packages/api-core/src/site-users/fetchers.ts:
```diff
@@ -14,3 +14,17 @@ export async function fetchSiteUsers( siteId: number ): Promise< SiteUser[] > {
 	} );
 }
+
+export async function fetchSiteUsersWpcom(
```
The naming seems a little inconsistent. Do you think we should call them "wpcom site users", or a site's "wpcom users"? It's just the order of the nouns that I'm nitpicking about :)

p-jackson commented on client/dashboard/domains/domain-transfer/transfer-domain-to-other-user.tsx:
```diff
@@ -32,7 +36,14 @@ export type TransferFormData = {
+
+const getUserDisplayName = ( user: WpcomSiteUser ) => {
+	const { first_name, last_name, nice_name } = user;
+	return first_name && last_name ? `${ first_name } ${ last_name } (${ nice_name })` : nice_name;
```
I don't think we can [simply concatenate first and last names like this](https://www.kalzumeus.com/2010/06/17/falsehoods-programmers-believe-about-names/). We have the "public display name" field for that sort of thing.

<img width="1172" height="1020" alt="CleanShot 2025-12-15 at 14 00 19@2x" src="https://github.com/user-attachments/assets/87f5d2f0-2fc1-41d8-8f89-3b066235fcb4" />

I'm not exactly sure what "nice name" is. But I think if we want to show the username to help disambiguate then it should something more like

```suggestion
	const { name, login } = user;
	return name ? `${ name } (${ login })` : login;
```

p-jackson commented on client/dashboard/domains/domain-transfer/transfer-domain-to-other-user.tsx:
```diff
@@ -60,7 +71,9 @@ const form = {
 	const { domainName } = domainRoute.useParams();
 	const { data: domain } = useSuspenseQuery( domainQuery( domainName ) );
-	const { data: users } = useSuspenseQuery( siteUsersQuery( domain.blog_id ) );
+	const { data: usersResponse } = useSuspenseQuery(
```
Since you're using React Suspense you will need to ensure the data is loaded in the router too

https://github.com/Automattic/wp-calypso/blob/1d46005688833af9a8e73961e1c65575d8b1b002/client/dashboard/app/router/domains.ts#L607

Otherwise it will flash an empty component while loading.

p-jackson commented on client/dashboard/domains/domain-transfer/transfer-domain-to-other-user.tsx:
```diff
@@ -41,8 +52,8 @@ const createFields = ( users: SiteUser[] ): Field< TransferFormData >[] => [
 						...users.map( ( user ) => ( {
-							value: user.id,
-							label: user.name,
+							value: getWpcomUserId( user ),
```
There's a mixture of strings and numbers being used as the element "value" (see line 53 where an empty string is being used).

I think "string" is what we would usually use, and so we should do `String( getWpcomUserId( user ) )` here. That way down on line 105 we're not doing the conversion there.

markbiek replied:
I went a slightly different direction that seemed more consistent with the the rest of that component. So now we're always comparing `int` to `int`.

p-jackson replied:
That seems fine. The `Choose an administrator on this site` item still users a string value, but I'm guessing the underlying HTML element doesn't really mind and probably coerces everything to a string anyway 🤷‍♂️

markbiek commented:
@p-jackson Thanks so much for that feedback. I somehow missed the notifications but will try and address those today!

markbiek committed:
Change query key for siteUsersWpcomQuery

markbiek committed:
Rename wpcom site users fetcher for clarity

markbiek committed:
Update transfer user pulldown

We user display_name or nice_name or email in that order. Previously we
tried concatenating first and last name and only falling back to
nice_name.

markbiek committed:
Clean up types when confirming transfer to correct user

markbiek committed:
Add usersResponse to the domain router

p-jackson commented on client/dashboard/domains/domain-transfer/transfer-domain-to-other-user.tsx:
```diff
@@ -89,15 +103,20 @@ export default function TransferDomainToOtherUser() {
 	const onConfirm = () => {
 		const selectedUser = availableUsers.find(
-			( user ) => ( user.id as unknown as string ) === formData.user
+			( user ) => getWpcomUserId( user ) === parseInt( formData.user )
```
Nit: I can't imagine the ID in the form data will ever end up with a leading `0` or anything, but I always feel its best practice to pass an explicit radix to `parseInt`

```suggestion
			( user ) => getWpcomUserId( user ) === parseInt( formData.user, 10 )
```

p-jackson commented on packages/api-queries/src/site-users.ts:
```diff
@@ -24,3 +29,9 @@ export const siteUsersQuery = ( siteId: number ) =>
+
+export const siteUsersWpcomQuery = ( siteId: number, role?: string ) =>
+	queryOptions( {
+		queryKey: [ 'site', siteId, 'wpcom-users', 'wpcom', role ],
```
Nit: I don't see us needing this extra layer in the query key, but it's not a huge deal

```suggestion
		queryKey: [ 'site', siteId, 'wpcom-users', role ],
```

Having `'wpcom'` there implies maybe there'd be other categories of `wpcom-users` in the future?

arthur791004 replied:
Question: Do we need `wpcom` here?

p-jackson reviewed (approved):
lgtm

arthur791004 commented on client/dashboard/app/router/domains.ts:
```diff
@@ -607,6 +608,7 @@ export const domainTransferToOtherUserRoute = createRoute( {
 	loader: async ( { params: { domainName } } ) => {
 		const domain = await queryClient.ensureQueryData( domainQuery( domainName ) );
 		await queryClient.ensureQueryData( domainTransferRequestQuery( domainName, domain.site_slug ) );
+		await queryClient.ensureQueryData( siteUsersWpcomQuery( domain.blog_id, 'administrator' ) );
```
Nit: I think we can request them together by `await Promise.all( [ ... ] )`

fushar replied:
Yes please!

arthur791004 commented on packages/api-core/src/site-users/fetchers.ts:
```diff
@@ -14,3 +14,17 @@ export async function fetchSiteUsers( siteId: number ): Promise< SiteUser[] > {
+
+export async function fetchWpcomSiteUsers(
+	siteId: number,
+	role?: string
```
Nit: Since role is optional here, would it be better to make it as something like `options` object with defined properties?

```js
options?: { role?: string }

...
query: {
  force: 'wpcom',
  ...options
}
```

arthur791004 commented on client/dashboard/domains/domain-transfer/transfer-domain-to-other-user.tsx:
```diff
@@ -188,7 +207,9 @@ export default function TransferDomainToOtherUser() {
 
 	const renderConfirmationDialog = () => {
-		const selectedUser = availableUsers.find( ( user ) => user.id.toString() === formData.user );
+		const selectedUser = availableUsers.find(
```
It seems we can move it to the component level so we don't need to have duplicate code to get selected user.

arthur791004 commented on packages/api-core/src/site-users/fetchers.ts:
```diff
@@ -14,3 +14,17 @@ export async function fetchSiteUsers( siteId: number ): Promise< SiteUser[] > {
+	siteId: number,
+	role?: string
+): Promise< WpcomSiteUsersResponse > {
+	return wpcom.req.get( {
```
Do we need the `found` property? If we don't have any use case before, maybe we can simply return `users` here to align the response of `fetchSiteUsers`

arthur791004 reviewed (approved):
The functionalities look good to me! Just left some minor feedback 🙂

markbiek commented:
> The functionalities look good to me! Just left some minor feedback 🙂

Thanks @arthur791004 ! I'll take a look at those today.

markbiek committed:
Use Promise.all for parallel query loading in domain transfer route

markbiek committed:
Remove redundant 'wpcom' from siteUsersWpcomQuery key

markbiek committed:
Use options object for fetchWpcomSiteUsers

markbiek committed:
Return users array directly from fetchWpcomSiteUsers

markbiek committed:
Extract selectedUser lookup to component-level useMemo

markbiek commented:
@arthur791004 I believe all of those are addressed if you'd like to take another look
